doc/topics/the-event-loop-timers-and-nexttick.md

```diff
@@ -0,0 +1,375 @@
+have to be.  Take this code snippet for example:
+
+```js
+function apiCall (arg, callback) {
+  if (typeof arg !== 'string')
+    return process.nextTick(callback,
+      new TypeError('argument should be string'));
+}
+```
+
+The snippet does an argument check and if it's not correct, it will pass the
+error to the callback.  The API updated fairly recently to allow passing
+arguments to `process.nextTick()` allowing it to take any arguments passed
+after the callback to be propagated as the arguments to the callback so you
+don't have to nest functions.
+
+What we're doing is passing an error back to the user but only *after* we have
+allowed the rest of the user's code to execute.  By using `process.nextTick()`
+we guarantee that `apiCall()` always runs its callback *after* the rest of the
+user's code and *before* the event loop is allowed to proceed.  To acheive
+this, the JS call stack is allowed to unwind then immediately execute the
+provided callback which allows a person to make recursive calls to nextTick
+without reaching a `RangeError: Maximum call stack size exceeded from v8`.
+
+This philosophy can lead to some potentially problematic situations.  Take
+this snippet for example:
+
+```js
+// this has an asynchronous signature, but calls callback synchronously
+function someAsyncApiCall (callback) { callback(); };
+
+// the callback is called before `someAsyncApiCall` completes.
+someAsyncApiCall(() => {
+
+  // since someAsyncApiCall has completed, bar hasn't been assigned any value
+  console.log('bar', bar); // undefined
+
+});
+
+var bar = 1;
+```
+
+The user defines `someAsyncApiCall()` to have an asynchronous signature,
+actually operates synchronously.  When it is called, the callback provided to
+`someAsyncApiCall ()` is called in the same phase of the event loop because
+`someAsyncApiCall()` doesn't actually do anything asynchronously.  As a
+result, the callback tries to reference `bar` but it may not have that
+variable in scope yet because the script has not been able to run to
+completion.
+
+By placing it in a `process.nextTick()`, the script still has the ability to
+run to completion, allowing all the variables, functions, etc., to be
+initialized prior to the callback being called.   It also has the advantage of
+not allowing the event loop to continue.  It may be useful that the user be
+alerted to an error before the event loop is allowed to continue.
+
+A real world example in node would be:
+
+```js
+const server = net.createServer(() => {}).listen(8080);
+
+server.on('listening', () => {});
+```
+
+When only a port is passed the port is bound immediately. So the `'listening'`
+callback could be called immediately. Problem is that the `.on
+('listening')` will
+not have been set by that time.
+
+To get around this the `'listening'` event is queued in a `nextTick()` to allow
+the script to run to completion. Which allows the user to set any event
+handlers they want.
+
+## process.nextTick() vs `setImmediate()`
```

Comment:
All `process.nextTick()` references should be surrounded by backticks consistently.